### PR TITLE
Fix hypernoblium formation gas reaction rate sanity clamping not accounting for the reduction factor. Also other sanity fixes.

### PR DESCRIPTION
**Pull request by Pickle-Coding**
## About The Pull Request
Fix hypernoblium formation not accounting for the reduction factor when sanity clamping the reaction rate.

Fix N2O formation not using updated consumption rates for sanity checks.

Fix BZ formation not accounting for nitrous oxide decomposition factor for reaction rate sanity clamping.
## Why It's Good For The Game
So the reactions are more natural.
## Changelog
:cl:
fix: Fix hypernoblium formation gas reaction rate sanity clamping not accounting for the reduction factor.
fix: Fix N2O formation gas reaction rate clamping clamping for incorrect consumption rates.
fix: Fix BZ formation reaction rate sanity clamping not accounting for N2O decomposition factor.
/:cl:

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 1]
```diff
@@ -448,8 +448,8 @@
 
 /datum/gas_reaction/nitrousformation/react(datum/gas_mixture/air)
 	var/list/cached_gases = air.gases
-	var/heat_efficency = min(cached_gases[/datum/gas/oxygen][MOLES], cached_gases[/datum/gas/nitrogen][MOLES] * INVERSE(2))
-	if ((cached_gases[/datum/gas/oxygen][MOLES] - heat_efficency < 0 ) || (cached_gases[/datum/gas/nitrogen][MOLES] - heat_efficency * 2 < 0))
+	var/heat_efficency = min(cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5), cached_gases[/datum/gas/nitrogen][MOLES])
+	if ((cached_gases[/datum/gas/oxygen][MOLES] - heat_efficency * 0.5 < 0 ) || (cached_gases[/datum/gas/nitrogen][MOLES] - heat_efficency < 0))
 		return NO_REACTION // Shouldn't produce gas from nothing.
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/reactions.dm:453]: undocced?

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 2]
```diff
@@ -779,14 +779,17 @@
 
 /datum/gas_reaction/nobliumformation/react(datum/gas_mixture/air)
 	var/list/cached_gases = air.gases
-	var/nob_formed = min((cached_gases[/datum/gas/nitrogen][MOLES] + cached_gases[/datum/gas/tritium][MOLES]) * 0.01, cached_gases[/datum/gas/tritium][MOLES] * INVERSE(5), cached_gases[/datum/gas/nitrogen][MOLES] * INVERSE(10))
-
-	if (nob_formed <= 0 || (cached_gases[/datum/gas/tritium][MOLES] - 5 * nob_formed < 0) || (cached_gases[/datum/gas/nitrogen][MOLES] - 10 * nob_formed < 0))
+	/// List of gases we will assert, and possibly garbage collect.
+	var/list/asserted_gases = list(/datum/gas/hypernoblium, /datum/gas/bz)
+	air.assert_gases(arglist(asserted_gases))
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/reactions.dm:784]: annoying

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 1 above)
Comment (Pickle-Coding) [reply code/modules/atmospherics/gasmixtures/reactions.dm:453]: Mobile copy paste failure moment.

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 2 above)
Comment (Pickle-Coding) [reply code/modules/atmospherics/gasmixtures/reactions.dm:784]: Guh?

File: code/modules/atmospherics/gasmixtures/reactions.dm (hunk 1 above)
Comment (Pickle-Coding) [reply code/modules/atmospherics/gasmixtures/reactions.dm:453]: Wait nevermind. What do you mean by undocced? What got undocced?

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/reactions.dm:453]: you hadn't mentioned this change in the desc when I got to it, so I got confused is all

**Comment by EliteCreature** [comment]
When you say 'Sanity Clamping', what does that mean? [Just my first time hearing this phrase]

**Comment by Pickle-Coding** [comment]
Clamp reaction rate to not consume more moles than there is, so no negative moles or creating gas from nothing bullshit.

**Pull request merged**